### PR TITLE
Ensure graphdriver only loads with experimental flag

### PR DESCRIPTION
**Pull request by cpuguy83**
Also updates some of the structures being sent so plugins are getting
all the new options.

ping @thaJeztah

### CODE REVIEW

File: docs/extend/plugins_graphdriver.md
Comment (thaJeztah) [inline docs/extend/plugins_graphdriver.md]: Forgot to remove "Experimental:" from the title

Comment (thaJeztah) [inline docs/extend/plugins_graphdriver.md]: Also, there should be metadata added for the docs (easiest to take other documents from the docker.github.io repo as example)

File: docs/extend/plugins_graphdriver.md
Comment (thaJeztah) [inline docs/extend/plugins_graphdriver.md]: Should be no trailing comma here

File: docs/extend/plugins_graphdriver.md
Comment (thaJeztah) [inline docs/extend/plugins_graphdriver.md]: Can you add a `json` language hint to the fence here, and for other JSON examples?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Left some comments

Also /cc @mstanleyjones because I think in the new docs, every page has to be added to the menu manually

File: docs/extend/plugins_graphdriver.md
Comment (AkihiroSuda) [inline docs/extend/plugins_graphdriver.md]: nit: we need a `,` in the end of this line

File: docs/extend/plugins_graphdriver.md
Comment (AkihiroSuda) [inline docs/extend/plugins_graphdriver.md]: ditto

File: docs/extend/plugins_graphdriver.md
Comment (AkihiroSuda) [inline docs/extend/plugins_graphdriver.md]: ditto

Comment (thaJeztah) [inline docs/extend/plugins_graphdriver.md]: comma here as well (thanks @AkihiroSuda )

File: docs/extend/plugins_graphdriver.md
Comment (thaJeztah) [inline docs/extend/plugins_graphdriver.md]: missing comma here as well

File: docs/extend/plugins_graphdriver.md
Comment (thaJeztah) [inline docs/extend/plugins_graphdriver.md]: missing comma

**Review by thaJeztah — CHANGES_REQUESTED** [review]
found some more missing commas

**Comment by cpuguy83** [comment]
This is updated with comma changes.

File: docs/extend/plugins_graphdriver.md
Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: What do you mean about metadata added? In the docs repo we will need to add a TOC reference for this file, but I don't understand what else you mean...

File: docs/extend/plugins_graphdriver.md
Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: This is an awkward wording. I don't really understand what it is talking about; Is there a driver back-end that has to be running before Docker Engine can be started?

File: docs/extend/plugins_graphdriver.md
Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: All of this passive voice makes this hard to understand. Maybe something like "The array of option is passed through to the plugin, but the plugin is not required to parse or honor them."

File: docs/extend/plugins_graphdriver.md
Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: How about something like:

The response also includes a list of UID and GID mappings, structured as follows:

File: docs/extend/plugins_graphdriver.md
Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: Maybe something like:

If `Parent` is an empty string, there is no parent layer.

`StoreOpt` is a map of strings which indicate storage options.

File: docs/extend/plugins_graphdriver.md
Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: s/for example/such as

File: docs/extend/plugins_graphdriver.md
Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: If `Parent` is an empty string, there is no parent.

**Review by mdlinville — CHANGES_REQUESTED** [review]

File: docs/extend/plugins_graphdriver.md
Comment (cpuguy83) [inline docs/extend/plugins_graphdriver.md]: I agree, I had trouble phrasing it... basically this:

```
$ docker plugin install myGraphDriver
$ docker plugin enable myGraphDriver
$ pkill dockerd
$ dockerd -s myGraphDriver
```

The daemon has to be started with the graphdriver that it is going to use. You can't use a graphdriver plugin unless the plugin is enabled before you restart the daemon.

Comment (mstanleyjones) [inline docs/extend/plugins_graphdriver.md]: How about "You need to install and enable the plugin and then restart the Docker daemon before using the plugin. See the following example for the correct ordering of steps." followed by your 4-line code block?

**Comment by cpuguy83** [comment]
Adding `[WIP]` because there's still a few kinks to work out.
Technically nothing is blocking someone from loading a plugin graphdriver on a non-experimental daemon, but moving docs over is another thing altogether.

**Comment by thaJeztah** [comment]
@cpuguy83 if it remains experimental, you can add `advisory: experimental` to the metadata of the docs, that way it will print a warning (but you can still have the documentation published on docs.docker.com)

**Comment by cpuguy83** [comment]
@thaJeztah Think I've worked out all the issues in any case, just blocked by #26398 at the moment.

**Review by thaJeztah — APPROVED** [review]
docs LGTM

**Comment by mstanleyjones** [comment]
Any thoughts on my suggestions above?

**Comment by cpuguy83** [comment]
@mstanleyjones 👍  to the suggestions. I'll get them in.

**Comment by thaJeztah** [comment]
@cpuguy83 https://github.com/docker/docker/pull/26398 was merged

**Comment by cpuguy83** [comment]
Ok, this is updated.
Gates loading a graphdriver via plugin by `--experimental` daemon flag... this is due to a few issues to deal with in the graphdriver interface, and makes sure we aren't stuck in the future because we need to support existing plugins.... and I doubt any pluggable graphdriver is really production ready at this point anyway.

ping @anusha-ragunathan

**Comment by cpuguy83** [comment]
ping @tiborvass @aaronlehmann

**Comment by tiborvass** [comment]
LGTM

File: daemon/graphdriver/plugin.go
Comment (anusha-ragunathan) [inline daemon/graphdriver/plugin.go]: I would add "graphdriver plugins are only..." just to be extra clear.

File: docs/extend/plugins_graphdriver.md
Comment (anusha-ragunathan) [inline docs/extend/plugins_graphdriver.md]: `dockerd --experimental -s ...`

File: integration-cli/docker_cli_daemon_plugins_test.go
Comment (anusha-ragunathan) [inline integration-cli/docker_cli_daemon_plugins_test.go]: Lets make the testRequires an `ExperimentalDaemon` instead. Easier to switch to stable, when its supported.

File: integration-cli/docker_cli_external_graphdriver_unix_test.go
Comment (anusha-ragunathan) [inline integration-cli/docker_cli_external_graphdriver_unix_test.go]: `testRequires(ExperimentalDaemon)` instead?

Comment (anusha-ragunathan) [inline integration-cli/docker_cli_external_graphdriver_unix_test.go]: Also, not sure why this test is affected?

Comment (anusha-ragunathan) [inline integration-cli/docker_cli_external_graphdriver_unix_test.go]: Actually, `setupTest` already sets up an `ExperimentalDaemon`. So this should not be necessary.

Comment (cpuguy83) [inline integration-cli/docker_cli_external_graphdriver_unix_test.go]: I think I ran into this because in order for this test to work for me, I have to manually run the make script (whereas I normally have a helper bash function) and I didn't set `DOCKER_EXPERIMENTAL=1` in doing so.
Looks good to remove this when setup properly ;)

**Comment by cpuguy83** [comment]
Updated

**Comment by anusha-ragunathan** [comment]
LGTM

**Comment by thaJeztah** [comment]
Windows CI failed because a Jenkins issue, so i restarted. I'm moving this to "merge", because it has two LGTM's, and I see misty's suggestions were addressed

**Pull request merged**
